lib/jackhammer/topic.rb

```diff
@@ -13,18 +17,45 @@ def subscribe_queues
     # for each message published.
     def publish(message, options)
       full_options = Jackhammer.configuration.publish_options.dup.merge options
-      @topic.publish message, full_options
+      topic.publish message, full_options
     end
 
     def queues
       return @queues if @queues
 
-      @queues = @queue_config.map do |name, options|
+      @queues = queue_config.map do |options|
         handler = MessageReceiver.new(options.delete('handler'))
-        routing = options.delete('routing_key') || fail(InvalidConfigError, "routing_key not found in #{options.inspect}")
+        routing = fetch_and_delete_key(options, ROUTING_KEY_KEY)
+        name = options.delete(QUEUE_NAME_KEY) || name_from_routing(routing)
         queue = Jackhammer.channel.queue(name, options)
-        Queue.new(topic: @topic, queue: queue, handler: handler, routing: routing)
+        Queue.new(topic: topic, queue: queue, handler: handler, routing: routing)
       end
     end
+
+    private
+
+    attr_reader :topic, :queue_config, :app_name
+
+    # `queue_config` can be either:
+    # - an array of options containing `queue_name` key
+    # or
+    # - a hash containing `queue_name => options` pairs
+    def normalize_queue_config(config)
+      return config if config.is_a?(Array)
+
+      config.map do |name, options|
+        options[QUEUE_NAME_KEY] = name
+        options
+      end
+    end
+
+    def fetch_and_delete_key(options, key)
+      options.delete(key) || fail(InvalidConfigError, "#{key} not found in #{options.inspect}")
+    end
+
+    def name_from_routing(routing)
+      fail(InvalidConfigError, "app_name must be set to determine queue name from routing") unless app_name
+      "#{app_name}.#{routing}"
```

Comment:
I would call it `routing_key` instead of general `routing` (it becomes `routing_key` anyway down the road in [Queue](https://github.com/RenoFi/jackhammer/blob/1cefee78a6418a14b34bca8ea01bdc3cdb14ff55/lib/jackhammer/queue.rb#L8))

Reply:
Renamed in https://github.com/RenoFi/jackhammer/pull/26/commits/46e1f10349fb550d45620ac6235d86e5f248f6c8